### PR TITLE
Packaging tools for x86

### PR DESCRIPTION
**Pull request by mexanick**
x86 packaging is done. After running `make rpm` two packages are created:

- `gem_reg_interface_extras-1.0.0-centos7.python2.7.x86_64.rpm`
- `xhal-xhalcore-1.0.0-centos7.x86_64.rpm`

## Description

These packages require `wiscrpcsvc`, `log4cplus` and `xercesc` (`xdaq` versions of latter two), `reg_interface-1.0.0-centos7.python2.7.x86_64.rpm`
Default installation is /opt/xhalcore/:

```bash
tree /opt/xhalcore/
/opt/xhalcore/
├── bin
│   └── gem_reg.py
├── include
├── lib
│   ├── librpcman.so
│   └── libxhal.so
└── scripts
```

Required env settings:
`LD_LIBRARY_PATH` should contain path to `libwiscrpcsvc.so` and `xhalcore/lib`
`GEM_ADDRESS_TABLE_PATH` should point to address table XML location. Address table file name should be `amc_address_table_top.xml`

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context

Proper packaging is required for seamless GEM operations 
## How Has This Been Tested?

Tested at `gem904daq01` with `eagle34` and `eagle64`
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by bdorney** [comment]
Naive question, what about an `slc6` rpm?  Or is this not going to be supported?

Additionally larger question:

> `GEM_ADDRESS_TABLE_PATH` should point to address table XML location. Address table file name should be `amc_address_table_top.xml`

This seems to conflict with what `setupMachine.sh` from `cmsgemos` is doing:

```
[gemuser@gem904qc8daq gemdaq]$ ll $GEM_ADDRESS_TABLE_PATH
total 40K
-rw-r-xr-x. 1 root root 6.6K May 11 19:33 connections.xml~
-rwxr-xr-x. 1 root root  981 May 14 09:51 linkuhaltables.sh
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link10.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link10.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link11.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link11.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link00.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link00.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link01.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link01.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link02.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link02.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link03.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link03.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link04.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link04.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link05.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link05.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link06.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link06.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link07.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link07.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link08.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link08.xml
lrwxrwxrwx. 1 root root   59 May 18 14:29 uhal_gem_amc_ctp7_link09.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_link09.xml
lrwxrwxrwx. 1 root root   56 May 18 14:31 uhal_gem_amc_ctp7_amc.xml -> /data/bigdisk/sw/address_table/uhal_gem_amc_ctp7_amc.xml
-rw-r--r--. 1 root root 6.6K May 30 09:07 connections.xml.backup~
-rw-r--r--. 1 root root 6.6K May 30 09:08 connections.xml.backup
-rw-r-xr-x. 1 root root 9.5K May 30 09:20 connections.xml
```

While a symlink of course could be made...it might be useful to use a unified schema between the two repo's.

File: python/ctp7/dacScan.py [hunk 1]
```diff
@@ -1,120 +0,0 @@
-#!/bin/env python
-"""
-Script to measure the DAC values on the VFATs
-By: Jared Sturdy (sturdy@cern.ch), Mykhailo Dalchenko (mykhailo.dalchenko@cern.ch)
-"""
```
Comment (bdorney) [inline python/ctp7/dacScan.py:5]: Is this script being included somewhere else?  Seems like a useful tool that should not be permanently deleted.

File: python/extras/makeDACtree.py [hunk 2]
```diff
@@ -1,57 +0,0 @@
...
-    for event in data[dactype]:
-      vfatN[0] = int(event[0])
-      dacinval[0] = int(event[1])
-      dacoutval[0] = int(event[2])
-      myT.Fill()
-    myT.AutoSave("SaveSelf")
-
-myF.cd()
-myT.Write()
-myF.Close()
```
Comment (bdorney) [inline python/extras/makeDACtree.py:56]: seems related to [this](https://github.com/cms-gem-daq-project/xhal/pull/72/files#r195073305) comment

File: python/reg_interface/gbt.py [hunk 3]
```diff
@@ -1,230 +0,0 @@
...
-        exit(os.EX_USAGE)
-
-    if ohSelect > 11:
-        printRed("The given OH index (%d) is out of range (must be 0-11)" % ohSelect)
-        exit(os.EX_USAGE)
-    if gbtSelect > 2:
-        printRed("The given GBT index (%d) is out of range (must be 0-2)" % gbtSelect)
-        exit(os.EX_USAGE)
-
-    programGBT(ohSelect, gbtSelect, command)
```
Comment (bdorney) [inline python/reg_interface/gbt.py:230]: This is a necessary script, especially for programming `GBTx`'s over the fiber link.  If it is being deleted I trust it's  being included elsewhere, if so where?

File: python/reg_interface/mcs.py [hunk 4]
```diff
@@ -1,79 +0,0 @@
...
-    return bytes
-
-def hex(number):
-    if number is None:
-        return 'None'
-    else:
-        return "{0:#0x}".format(number)
-
-if __name__ == '__main__':
-    main()
```
Comment (bdorney) [inline python/reg_interface/mcs.py:79]: Again, where is this script being moved too?

File: python/reg_interface/rw_reg.py [hunk 5]
```diff
@@ -1,372 +0,0 @@
...
-    ret = string
-    for varKey in vars.keys():
-        ret = ret.replace('${' + varKey + '}', str(vars[varKey]))
-    return ret
-
-def tabPad(s,maxlen):
-    return s+"\t"*((8*maxlen-len(s)-1)/8+1) 
-
-if __name__ == '__main__':
-    main()
```
Comment (bdorney) [inline python/reg_interface/rw_reg.py:372]: Right now this is necessary for all `v3` actions.  Where will this file be found at in the future? e.g. what is the proper import.

File: python/reg_interface/test.py [hunk 6]
```diff
@@ -1,97 +0,0 @@
...
-            pass
-
-    return
-
-if __name__ == '__main__':
-  start_time = timeit.default_timer()
-  # code you want to evaluate
-  main()
-  elapsed = timeit.default_timer() - start_time
-  print "Total time %s" %(elapsed)
```
Comment (bdorney) [inline python/reg_interface/test.py:97]: same comment, where is this getting moved too?

**Review by bdorney — CHANGES_REQUESTED** [review]
Clarification requested on several scripts being deleted from the repo and request to change `amc_address_table_top.xml` to something that sync's better with `cmsgemos`.

**Comment by jsturdy** [comment]
@bdorney `reg_interface` "basic" and friends are now in [`reg_utils`](https://github.com/cms-gem-daq-project/reg_utils/tree/master/python/reg_interface), with the "GEM specific" extensions provided by `xhal`

> This seems to conflict with what setupMachine.sh from cmsgemos is doing:

`setupMachine.sh` doesn't do anything related to package structure, this is taken care of by the machine profile RPM, e.g., `cmsgemos-gem904qc8` (currently incomplete).  Though it currently lives in `cmsgemos` it will migrate to the "meta" package repository to properly include all external inputs (`GEM_AMC` generated address tables, TCDS configurations, setup specific patches to certain `python` files, etc.) as needed (whether through a script, or as an RPM installing a fixed "meta" release).  Currently installing the machine profile will create the structure, and provide a few setup specific files, but the address tables are currently updated manually.

**Review by jsturdy — COMMENTED** [review]
Not sure I understand the `xhalcore` division, is it a placeholder for something future?

File: python/reg_interface/gbt.py (hunk 3 above)
Comment (mexanick) [reply python/reg_interface/gbt.py:230]: it is for the moment under `reg_utils`. Maybe we will move it back to gem-specific extension. I'm still waiting inputs from @evka85

File: python/ctp7/dacScan.py (hunk 1 above)
Comment (mexanick) [reply python/ctp7/dacScan.py:5]: No, it is missing. I thought it was migrated under `gempython` - is it not?

File: python/reg_interface/mcs.py (hunk 4 above)
Comment (mexanick) [reply python/reg_interface/mcs.py:79]: Again, it is under `reg_utils`

File: python/ctp7/dacScan.py (hunk 1 above)
Comment (jsturdy) [reply python/ctp7/dacScan.py:5]: it's in a private branch (was not intended for public consumption even when i gave it to you :-) )

File: python/reg_interface/rw_reg.py (hunk 5 above)
Comment (mexanick) [reply python/reg_interface/rw_reg.py:372]: It depend on what you want to import. The functionality is divided in a following way:
`reg_utils.reg_interface.common.reg_xml_parser` takes care about XML parsing-related business like `parseXML` method and contains `Node` class definition
`reg_utils.reg_interface.common.reg_base_ops` loads a very basic `librwreg.so` 
`xhal.gem_reg_interface_extras.core.reg_extra_ops` provides access to `librpcman.so` functions

File: python/reg_interface/test.py (hunk 6 above)
Comment (mexanick) [reply python/reg_interface/test.py:97]: for the moment nowhere. Will be added later

**Comment by mexanick** [comment]
Certain headers may be added later under `xhalcore/include`. Testing will be added eventually under scripts. What else is unclear?

**Comment by mexanick** [comment]
Considering the address table: for the moment `setupMachine` seems to not care about address tables supplied by @evka85 and @andrewpeck at all. I would prefer to have a generic name there because the XML parser is a part of generic, non gem-specific `reg_utils/reg_interface` package

**Comment by jsturdy** [comment]
> Certain headers may be added later under xhalcore/include. Testing will be added eventually under scripts. What else is unclear?

I guess my question was more, why is the package called `xhalcore` rather than `xhal`?
Installed into a location `/opt/xhal`, and if sub packages exist at some point, then package them as `xhal-<subpackage>`, e.g., `xhal-core`

**Comment by mexanick** [comment]
ah, that's because of some name conventions and makefiles... I can try to revisit it if you insist

**Comment by jsturdy** [comment]
I have a moderate preference for revisiting :-), as I find the `xhalcore` awkward

**Comment by mexanick** [comment]
namespace changed to `xhal`
@bdorney anything you still want to change? If so, please elaborate :)

File: python/reg_interface/gbt.py (hunk 3 above)
Comment (bdorney) [reply python/reg_interface/gbt.py:230]: Okay so we are just moving the location, not removing it.  Got it.

File: python/reg_interface/mcs.py (hunk 4 above)
Comment (bdorney) [reply python/reg_interface/mcs.py:79]: Understood.

File: python/reg_interface/rw_reg.py (hunk 5 above)
Comment (bdorney) [reply python/reg_interface/rw_reg.py:372]: Okay I understood.  So some adding of multiple imports will resolve any dependency.

**Comment by bdorney** [comment]
Okay so it seems my concerns are just my own ignorance!  I'll need to consult the updated documentation for `xhal` and `reg_utils`.

**Review by bdorney — APPROVED** [review]

**Pull request merged**